app/tui/controller.py

```diff
@@ -0,0 +1,345 @@
+
+        current_index = available_indices.index(self.selected_process.index)
+        new_index = ((current_index + direction) % len(self.filtered_process_list))
+        self._tui_state.set_selected_process_by_index(new_index)
+
+    def sidebar_up(self):
+        self.move(-1)
+
+    def sidebar_down(self):
+        self.move(1)
+
+    def start_filter(self):
+        logger.info('in start_filter')
+        self._tui_state.filter_mode = True
+        self._tui_state.apply_filter('')
+        self._tui_state.selected_process = None
+        self.focused_widget = FocusWidget.SIDE_BAR_FILTER
+        self.refresh_app()
+
+    def update_filter(self, buffer: Buffer):
+        logger.info(f'in update filter: {buffer.text}')
+        self._tui_state.apply_filter(buffer.text)
+
+    def exit_filter(self, search_text: str):
+        logger.info('in exit_filter')
+        self._tui_state.apply_filter(search_text)
+        self._tui_state.filter_mode = False
+        self.focus_to_sidebar()
+
+    def view_docs(self):
+        logger.info('in view_docs')
+        self.toggle_docs()
+
+    def close_docs(self):
+        logger.info('in close_docs')
+        self.toggle_docs()
+
+    def toggle_docs(self):
+        logger.info(f'setting docks_open to {not self.docs_open}')
+        self._tui_state.docs_open = not self.docs_open
+        if self.docs_open:
+            self.focused_widget = FocusWidget.DOCS
+        else:
+            self.focused_widget = FocusWidget.SIDE_BAR_SELECT
+
+    def switch_focus(self):
+        logger.info('in switch_focus')
+        if not self.is_focused_on_free_form_input:
+            self.toggle_sidebar_terminal_focus()
+
+    def zoom(self):
+        logger.info('in zoom')
+        if not self.is_focused_on_free_form_input:
+            self.toggle_zoom()
+
+    def toggle_zoom(self):
+        logger.info('in zoom')
+        if self.zoomed_in:
+            logger.info('setting zoomed_in to False')
+            self._tui_state.zoomed_in = False
+            self.focus_to_sidebar()
+        else:
+            if self.focus_to_current_terminal():
+                logger.info('setting zoomed_in to True')
+                self._tui_state.zoomed_in = True
+
+    # /Sidebar
+
+    # Scroll
+
+    def register_scroll_mode_change_handler(self, handler: Callable[[int, bool], None]):
+        self._on_scroll_mode_changed_handlers.append(handler)
+
+    def on_scroll_mode_change(self, process: Process):
+        for handler in self._on_scroll_mode_changed_handlers:
+            handler(process.index, process.scroll_mode)
+
+    def toggle_scroll(self):
+        logger.info('in _toggle_scroll')
+        if not self.is_focused_on_free_form_input and self.selected_process:
+            self.selected_process.scroll_mode = not self.selected_process.scroll_mode
+            self.on_scroll_mode_change(self.selected_process)
+            if not self.selected_process.scroll_mode:
+                self.focus_to_sidebar()
+
+    # /Scroll
```

Comment:
i love how clean and discrete each of these methods are. It will be nice that the callers of the controller class have all of these actions at their disposal.
 
It seems like each of the category comments like "#scroll", "#sidbar" etc delineate nice chunks of logic that could be refactored to a new class. Then we could use the main controller class as kind of a facade pattern where we aggregate all of the available actions and expose them as a single class/module. 
wdyt?

Reply:
I was thinking the same thing, wasn't going to tackle it as part of this of this effort, but would definitely be a nice improvement. I feel controller is already a little too big, and it will just keep growing. The new CardEngine